### PR TITLE
tcp_proxy: support command operators in tunneling_config.hostname

### LINKED ISSUE #19612: Dynamic setting value of TcpProxy.tunneling_config.hostname
*Description*:
I have a case where applications running in Kubernetes communicate with external services (outside Kubernetes and in another network) through a dedicated forward proxy. Apps send HTTP CONNECT request and the proxy opens a TCP tunnel. The flow is illustrated below.
scenario

I would like to make forward proxy transparent for apps by routing requests from applications via Envoy to forward proxy. This would allow to avoid sending HTTP CONNECT requests from apps. The desired solution is illustrated below.
solution

It's quite easy to achieve the desired flow by creating the following listener:
```
static_resources:
  listeners:
  - name: egress_gateway_listener
    address:
      socket_address:
        address: 0.0.0.0
        port_value: 443
    listener_filters:
    - name: envoy.filters.listener.tls_inspector
    filter_chains:
    - filters:
      - name: tcp
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
          cluster: external_proxy_cluster
          tunneling_config:
            hostname: example.com:443
  clusters:
  - name: external_proxy_cluster
    connect_timeout: 0.25s
    type: strict_dns
    lb_policy: round_robin
    load_assignment:
      cluster_name: proxy_cluster
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: external-proxy
                port_value: 3128
```

The problem is that the `tunneling_config.hostname` can only be statically defined, so I can't tunnel traffic to multiple destinations having a single listener. One of the potential solutions would be to set `hostname` based on SNI. TLS Inspector filter provides SNI value in `%REQUESTED_SERVER_NAME%`, but it's not possible to overwrite `Host` header.
```
tunneling_config:
  hostname: example.com:443
  headers_to_add:
  - header:
      key: Host
      value: "%REQUESTED_SERVER_NAME%"
    append_action: OVERWRITE_IF_EXISTS_OR_ADD
```

So what do you think about to make `Host` header mutable or to figure out another way to dynamically set `tunneling_config.hostname`?

*Related issues*
#13809 #18838

### PR DESCRIPTION
**Pull request by jewertow**
Signed-off-by: Jacek Ewertowski <jacek.ewertowski1@gmail.com>

Commit Message: tcp_proxy: support command operators in tunneling_config.hostname
Additional Description: This change enables dynamically setting `tunneling_config.hostname` with command operators.
Risk Level: Low
Testing: added unit tests
Docs Changes: done
Release Notes: done
Platform Specific Features: none
[Optional Fixes #19612 #21804]

This pull request is an alternative for [auto_sni](https://github.com/envoyproxy/envoy/pull/20230).

This change allows to configure TCP proxy as follows:
```
tunneling_config:
  hostname: %REQUESTED_SERVER_NAME%:443
```

### CODE REVIEW

**Comment by jewertow** [comment]
@ramaraochavali @lambdai

**Review by ramaraochavali — COMMENTED** [review]
Thank you. LGTM

**Comment by jewertow** [comment]
@alyssawilk could you take a look at this pull request? I didn't add integration test yet, but you can review the general idea and implementation. This approach makes it possible to support also other [command operators](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#command-operators) if make sense. The advantage of this approach in contrary to overwriting `Host` header is that the `hostname` is not empty nor set to a dummy value what would be confusing.

I gave up the idea of `auto_sni`, because it's not backward compatible to wrap `hostname` and `auto_sni` in `oneof`. So the idea does not make sense, because it would be required to define empty or dummy hostname and I wanted to avoid it.

File: source/common/tcp_proxy/tcp_proxy.h
Comment (kyessenov) [inline source/common/tcp_proxy/tcp_proxy.h:127]: I wonder why we don't just support the entire formatter dictionary here? It's a bit strange and error-prone to just pick one variable.

Comment (alyssawilk) [reply source/common/tcp_proxy/tcp_proxy.h:127]: yeah it's great to see this hanging together end to end, but the implementation seems pretty fragile.  I'd suggest a standalone utility to swap the host part of the host-part pair, 
Also maybe a string_view for requested_server_name and a utility to handle the optional deref rather than passing all the downstream info through, WDYT?
/wait

Comment (jewertow) [reply source/common/tcp_proxy/tcp_proxy.h:127]: I don't see a need to support values like: `%START_TIME%` or `%UPSTREAM_PROTOCOL%`. Why would anyone need other values from dictionary?

Comment (kyessenov) [reply source/common/tcp_proxy/tcp_proxy.h:127]: @jewertow Yes, those are not applicable, but I think headers, things like metadata, filter state, or other TLS attributes (DNS SAN) are reasonable.

Comment (jewertow) [reply source/common/tcp_proxy/tcp_proxy.h:127]: Unfortunately headers couldn't be used, because it's a TcpProxy, not an HttpConnectionManager, so these data are not available. When it comes to TLS attributes, there are only available SANs of downstream proxy (`%DOWNSTREAM_LOCAL_URI_SAN%`, `%DOWNSTREAM_PEER_URI_SAN%`) and using these values does not make sense, because it would connect back to the downstream - I can't imagine any real use case. Other TLS attributes don't make sense in the context of hostname.

I agree that dynamic metadata or filter state might somehow be useful to provide hostname, but only these two. I considered all other command operators and can't imagine that any of them would be useful in the context of hostname. So I don't agree that supporting all command operators is important, because almost all of them are useless and it's not worth to implement it, test all cases and then maintain useless features and complex code base.

> It's a bit strange and error-prone to just pick one variable.

I don't agree that it's error-prone, because passing a command operator which is evaluated to irrelevant value is also strange error-prone. I'm going to document it, so it won't be confusing nor strange.

I understand your motivation and respect your feedback, but I just don't believe in this idea and I really think it's useless.
Dynamic metadata and filter state might be somehow useful, but I suggest to not support them as far as users don't need them. This PR aims to solve a real and common problem, so I am not convinced to extend its scope if it's not needed.

File: source/common/tcp_proxy/tcp_proxy.h
Comment (kyessenov) [inline source/common/tcp_proxy/tcp_proxy.h:124]: Does this throw an exception when `hostname_` has no `:`? I think we allowed hosts without full port form before.
Also, could we avoid parsing it every time? Would make more sense to only parse once and reject bad config rather than throwing on data path.

File: source/common/tcp_proxy/tcp_proxy.h
Comment (kyessenov) [reply source/common/tcp_proxy/tcp_proxy.h]: I am not disagreeing about the value or that it only makes sense with a subset of formatter strings. I am trying to avoid having special custom parsing code since it's difficult to maintain. I'm OK just supporting SNI for now, and tackle metadata/filter state later. As I mentioned earlier, we'll have to support URL templates for MASQUE which need generic templates.

**Review by kyessenov — COMMENTED** [review]
Please make sure exceptions are not thrown on the data path, otherwise, looks reasonable.

File: source/common/tcp_proxy/tcp_proxy.h
Comment (kyessenov) [reply source/common/tcp_proxy/tcp_proxy.h:124]: Checking on the spec: I think CONNECT requires `:` but it's no longer required in extended CONNECT. I also didn't find any place where it's enforced, so this probably deserves better validation on configuration.

Comment (kyessenov) [reply source/common/tcp_proxy/tcp_proxy.h:124]: Extended CONNECT was added later to h2 to support websockets and other non-TCP payloads: https://datatracker.ietf.org/doc/html/rfc8441.

**Comment by KBaichoo** [comment]
@kyessenov I think this is waiting on your reply; PTAL

@jewertow seems like you need to merge main

**Comment by lambdai** [comment]
Ping @jewertow 
Are you still working on it? It's an easy one to remove the exception. 

I'd like to propose the richer formatter on top of this change. Let me know if you want to complete this one

File: test/common/tcp_proxy/upstream_test.cc
Comment (howardjohn) [inline test/common/tcp_proxy/upstream_test.cc:394]: Can we support arbitrary header additions? See https://github.com/envoyproxy/envoy/issues/21804. This is a major blocker for us after some recent changes landed. Given we can add arbitrary metadata for all headers _except_ `:authority` today it seems like this should be feasible?

Comment (jewertow) [reply test/common/tcp_proxy/upstream_test.cc:394]: This change aims to allow setting command operators in the `tunneling_config.hostname` field to avoid setting empty or dummy hostname and then overriding header "Host" or ":authority" which are not allowed to override. But as @kyessenov requested, I'm refactoring this implementation to rely on substitution formatter which supports `%DYNAMIC_METADATA%`, so this change should provide what you need.

File: test/common/tcp_proxy/upstream_test.cc
Comment (jewertow) [reply test/common/tcp_proxy/upstream_test.cc:438]: cc: @lambdai

Comment (howardjohn) [reply test/common/tcp_proxy/upstream_test.cc:438]: awesome, thanks!

**Comment by jewertow** [comment]
@kyessenov I implemented this feature with substitution formatter as you requested and added a unit test for `DYNAMIC_METADATA`. Could you review once again?

File: source/common/tcp_proxy/tcp_proxy.h
Comment (kyessenov) [inline source/common/tcp_proxy/tcp_proxy.h:134]: This is uninitialized. You should store by value or better use `Http::StaticEmptyHeaders`.

File: envoy/tcp/upstream.h
Comment (jewertow) [reply envoy/tcp/upstream.h]: Done

File: source/common/tcp_proxy/tcp_proxy.cc
Comment (jewertow) [reply source/common/tcp_proxy/tcp_proxy.cc]: Done

File: source/common/tcp_proxy/tcp_proxy.cc
Comment (jewertow) [reply source/common/tcp_proxy/tcp_proxy.cc]: I added comments to proto and briefly described the change in the current changelog.

File: source/common/tcp_proxy/tcp_proxy.cc
Comment (jewertow) [reply source/common/tcp_proxy/tcp_proxy.cc]: Fixed

File: source/common/tcp_proxy/tcp_proxy.h
Comment (jewertow) [reply source/common/tcp_proxy/tcp_proxy.h]: Fixed

File: changelogs/current.yaml
Comment (kyessenov) [inline changelogs/current.yaml:37]: I think the label should be:
`envoy_v3_api_field_extensions.filters.network.tcp_proxy.v3.TcpProxy.TunnelingConfig.hostname`

File: source/common/tcp_proxy/tcp_proxy.h
Comment (kyessenov) [inline source/common/tcp_proxy/tcp_proxy.h:136]: Make this `const Formatter::FormatterPtr`.

File: source/common/tcp_proxy/tcp_proxy.cc
Comment (kyessenov) [inline source/common/tcp_proxy/tcp_proxy.cc:82]: I think this code could be made part of `TunnelingConfigHelperImpl`. E.g. add a function there, and initialize const formatter field with a function call.

File: test/common/tcp_proxy/upstream_test.cc
Comment (kyessenov) [inline test/common/tcp_proxy/upstream_test.cc:51]: If you moved formatter construction into helper you wouldn't have to duplicate this logic several times.

**Review by kyessenov — CHANGES_REQUESTED** [review]
Generally looks good, but there are a few comments that could improve this I think.

**Comment by kyessenov** [comment]
@jewertow Please don't force push. That makes it hard to track previous comments. Thanks.
/wait

File: source/common/tcp_proxy/tcp_proxy.h
Comment (jewertow) [reply source/common/tcp_proxy/tcp_proxy.h:141]: I can't, because `fromProtoConfig` does not return `const`.

File: changelogs/current.yaml
Comment (jewertow) [reply changelogs/current.yaml]: Thanks, I fixed it.

File: source/common/tcp_proxy/tcp_proxy.cc
Comment (jewertow) [reply source/common/tcp_proxy/tcp_proxy.cc]: Good idea, I did it.

File: test/common/tcp_proxy/upstream_test.cc
Comment (jewertow) [reply test/common/tcp_proxy/upstream_test.cc]: Done

File: test/common/tcp_proxy/upstream_test.cc
Comment (kyessenov) [inline test/common/tcp_proxy/upstream_test.cc:218]: nit: these could be `const std::string&`

File: source/common/tcp_proxy/tcp_proxy.h
Comment (kyessenov) [inline source/common/tcp_proxy/tcp_proxy.h:119]: nit: It's better to move this implementation code to `.cc` file to prevent header pollution. also `this->` is unusual, but not a big deal.

**Review by kyessenov — APPROVED** [review]
LGTM, need @envoyproxy/api-shepherds approval for proto change, and a couple of code nits.

File: source/common/tcp_proxy/tcp_proxy.h
Comment (jewertow) [reply source/common/tcp_proxy/tcp_proxy.h:119]: Ok, I moved constructor and `host` method to .cc file.

File: test/common/tcp_proxy/upstream_test.cc
Comment (jewertow) [reply test/common/tcp_proxy/upstream_test.cc]: Done

**Review by kyessenov — APPROVED** [review]

File: source/common/tcp_proxy/tcp_proxy.h
Comment (lizan) [reply source/common/tcp_proxy/tcp_proxy.h:128]: You can if you initialize this in initialization list and extract
```
envoy::config::core::v3::SubstitutionFormatString substitution_format_config;
  substitution_format_config.mutable_text_format_source()->set_inline_string(
      config_message.hostname());
  return Formatter::SubstitutionFormatStringUtils::fromProtoConfig(
      substitution_format_config, context);
```
to a helper function.

**Review by lizan — COMMENTED** [review]
/lgtm api

**Pull request merged**
